keps/sig-storage/1979-object-storage-support/README.md

```diff
@@ -73,636 +85,560 @@
+   1. Controller removes `objectstorage.k8s.io/has-bucketaccess-references` from BucketClaim if this is the last BucketAccess against the BucketClaim (this allows BucketClaim to start deletion, if applicable)
+   2. Controller removes `objectstorage.k8s.io/bucketaccess-protection` from BucketAccess
+
+#### Attaching Bucket Information to Pods
 
-## Bucket
+User attaches bucket information to their workload's application pod.
+
+1. User references the BucketAccess secret using the pod volume downward API in their pod spec
+2. User configures pod container(s) to mount secret data items as desired
+
+See [BucketAccess Secret data](#bucketaccess-secret-data) section for more details.
+
+The BucketAccess secret can be provided to the pod using any Kubernetes {Secret -> Pod} attachment mechanism. This naturally includes mounting data into environment variables and files. Mounting credential data into files is slightly more secure than environment variables and is thus recommended. However, each application has different requirements, and some may require environment variables for configuring access.
+
+### Sharing Buckets
+
+This section describes the current design for sharing buckets with other namespaces. As of COSI v1alpha2, any BucketClaims created in one namespace cannot be accessed in another namespace. I.e. no bucket sharing is possible. In future versions, a namespace-level access control will be enforced, and Buckets will be constrained to particular namespaces using selectors. Admins will be able to control which namespaces can access which buckets using namespace selectors.
+
+### COSI API Reference
+
+<!-- TODO: clarify how to reuse access for multiple buckets -->
+
+#### Annotations and finalizers
+
+Annotations:
+- `objectstorage.k8s.io/bucketclaim-being-deleted`: applied to a Bucket when the Controller detects that the Bucket's bound BucketClaim is being deleted
+- `objectstorage.k8s.io/has-bucketaccess-references`: applied to a BucketClaim when the Controller detects that one or more BucketAccesses reference the claim
+- `objectstorage.k8s.io/sidecar-cleanup-finished`: applied to a BucketAccess when the Sidecar has finished cleaning up, allowing the Controller to begin its final cleanup operations
+
+Finalizers:
+- `objectstorage.k8s.io/bucketclaim-protection`: applied to BucketClaims to prevent them from being deleted until COSI has cleaned up intermediate and underlying resources
+- `objectstorage.k8s.io/bucket-protection`: applied to Buckets to prevent them from being deleted until COSI has cleaned up underlying resources
+- `objectstorage.k8s.io/bucketaccess-protection`: applied to BucketAccesses and BucketAccess Secrets to prevent them from being deleted until COSI has cleaned up underlying resources
+
+#### Bucket
 
-Resource to represent a Bucket in OSP. Buckets are cluster-scoped.
+Resource to represent a Bucket in an OSP. Bucket is cluster-scoped.
 
 ```go
 Bucket {
   TypeMeta
   ObjectMeta
 
   Spec BucketSpec {
```

Comment:
One more implementation nitpick: we should follow standard CRD format with spec field, instead of inlining the specification to the resources. Eg. `(BucketClass).spec.params` instead of `(BucketClass).params`;

Reply:
Are you referring to the fact that BucketClass and AccessClass don't use `Spec` in the KEP? I missed that detail, but I do agree that we should use `Spec` which is standard.